storm-client/src/jvm/org/apache/storm/windowing/persistence/WindowState.java

```diff
@@ -0,0 +1,422 @@
+                                deletePartition(pid);
+                                windowPartitionsState.delete(pid);
+                            }
+                        } else if (p.isModified()) {
+                            windowPartitionsState.put(pid, p);
+                        } else {
+                            LOG.debug("WindowPartition '{}' is not modified", pid);
+                        }
+                    } finally {
+                        windowPartitionsLock.unlock(pid);
+                    }
+                }
+            }).build(new WindowPartitionCache.CacheLoader<Long, WindowPartition<T>>() {
+                @Override
+                public WindowPartition<T> load(Long id) {
+                    LOG.debug("Load partition: {}", id);
+                    // load from state
+                    try {
+                        windowPartitionsLock.lock(id);
+                        return windowPartitionsState.get(id, new WindowPartition<>(id));
+                    } finally {
+                        windowPartitionsLock.unlock(id);
+                    }
+                }
+            });
+    }
+
+    private void deletePartition(long pid) {
+        LOG.debug("Delete partition: {}", pid);
+        try {
+            partitionIdsLock.lock();
+            partitionIds.remove(pid);
+            partitionIdsState.put(PARTITION_IDS_KEY, partitionIds);
+        } finally {
+            partitionIdsLock.unlock();
+        }
+    }
+
+    private long getNextPartitionId() {
+        try {
+            partitionIdsLock.lock();
+            partitionIds.add(++latestPartitionId);
+            partitionIdsState.put(PARTITION_IDS_KEY, partitionIds);
+        } finally {
+            partitionIdsLock.unlock();
+        }
+        return latestPartitionId;
+    }
+
+    private WindowPartition<T> getPinnedPartition(long id) {
+        return cache.pinAndGet(id);
+    }
+
+    private void flush() {
+        LOG.debug("Flushing modified partitions");
+        cache.asMap().forEach((pid, p) -> {
+            Long pidToInvalidate = null;
+            try {
+                windowPartitionsLock.lock(pid);
+                if (p.isEmpty() && pid != latestPartitionId) {
+                    LOG.debug("Invalidating empty partition {}", pid);
+                    deletePartition(pid);
+                    windowPartitionsState.delete(pid);
+                    pidToInvalidate = pid;
+                } else if (p.isModified()) {
+                    LOG.debug("Updating modified partition {}", pid);
+                    p.clearModified();
+                    windowPartitionsState.put(pid, p);
+                }
+            } finally {
+                windowPartitionsLock.unlock(pid);
+            }
+            // invalidate after releasing the lock
+            if (pidToInvalidate != null) {
+                cache.invalidate(pidToInvalidate);
```

Comment:
it will be removed when (1) that partition gets unpinned and the entry is evicted from the cache due to loading other entries or (2) the next time the flush happens and the partition is empty.